### DISCUSSION TITLE
Request information on Xaml Controls Gallery (XCG) app resizing

### DISCUSSION THREAD

**Discussion by rcohn** [microsoft/WinUI-Gallery#839] [Q&A] [answered]
Hello -

I am trying to write a WinUI 3 app using Winrt/C++ and using the XCG as a paradigm.  One thing I'm still trying to understand is how the XCG elements wrap and prevent the app window from being resized below the point at which elements would be clipped.

I know that a VisualStateManager is used in most xaml pages to specify breakpoint triggers (eg: Breakpoint640Plus), which can result in actions being taken (such as rearranging elements as a window is resized smaller).  But I don't understand what prevents the apps from resizing below a certain width (or height).  Is there somebody on your team that can explain this or point out places in the code I can look for answers?

Thank you.

**Comment by gregwoo-microsoft** [comment, marked as answer]
Hi @rcohn! Thanks for your question. You should have a look at the [Window resizing sample](https://docs.microsoft.com/en-us/samples/microsoft/windows-universal-samples/resizeappview/). It shows you how you can set the minimum size of your app using `ApplicationView.SetPreferredMinSize(new Size { Width = 300, Height = 200 });` The docs for this API can be found [here](https://docs.microsoft.com/en-us/uwp/api/windows.ui.viewmanagement.applicationview.setpreferredminsize?view=winrt-22000). I hope this answers your question!

**Reply by rcohn** [reply]
Hi @gregwoo-microsoft -
Thanks so much for your reply.  I am familiar with the SetPreferredMinSize() method, but there seems to be a problem in using it within a WinUI3 Winrt/Cpp desktop app - or I am not calling it at the correct point in the app startup.  I believe (or at least it seems) that this method works correctly in UWP apps (as the ResizeAppView app is), but may have problems in the WinUI3 desktop app.  

Specifically, when I call SetPreferredMinSize() from the app's OnLaunch() method, it always crashes.  Here is a code snippet:
```
void App::OnLaunched(LaunchActivatedEventArgs const& e)
{
   HWND hwnd;
   window = make<MainWindow>();
   window.as<IWindowNative>()->get_WindowHandle(&hwnd);
   window.Activate();

   ApplicationView::GetForCurrentView().SetPreferredMinSize(Size(490, 590));
}
```
Apparently, the current view has not yet been created?  If I am using SetPreferredMinSize() from the wrong place, then can you advise me on the correct way to do it?

Again, my thanks.

**Reply by gregwoo-microsoft** [reply]
You are right, this solution is for UWP apps. I'll reach out to my team to see what should be the ideal way to do this.

**Reply by marb2000** [reply]
There are several UWP APIs that don't work on WinUI3 and many of them follow the pattern GetForCurrentView. There are docs (still in development) about how to migrate from UWP to WinUI3 (e.g, take a look at: https://docs.microsoft.com/en-us/windows/apps/windows-app-sdk/migrate-to-windows-app-sdk/guides/windowing).

About Setting the min size of a Window you need to use the Win32 APIs given right now there is no equivalent in the Windows App SDK 1.0 . Unfortunately, it's not easy. I wrote a sample in C# several months ago to demonstrate it: https://github.com/marb2000/DesktopWindow. Basically you need to [subclass](https://docs.microsoft.com/en-us/windows/win32/controls/subclassing-overview) the Window (HWND) to hook to the WinProc and do whatever you want (like listening to the [WM_GETMINMAXINFO](https://github.com/marb2000/DesktopWindow/blob/abb21b797767bb24da09c066514117d5f1aabd75/WindowExtensions/DesktopWindow.cs#L338) message to set the min/max size of a Window), and then call the original WinProc.
But I'm pretty sure that your code will be cleaner than mine given you don't need Pinvokes and marshallings in C++. 😊

**Reply by rcohn** [reply]
Hi @marb2000 -

Thanks very much for your detailed reply.  The WindowExtensions code should be quite helpful (although porting to c++ will take some doing).  It's interesting that the WinUI3 framework at its current 1.0 release still appears to be pretty much in the same state (as far as the need to do specialized window management as you have provided) as when you created this for ProjectReunion 0.5.

Best regards.

**Reply by gregwoo-microsoft** [reply]
I've marked your question as "answered" based on @marb2000's comment. Of course, feel free to let us know if there is something else we can help with :)
